File: unionFind.go

// example
func unionFind() {
        parent := make([]int, n)
	for i := 0; i < len(parent); i++ {
		parent[i] = i
	}
	rank := make([]int, n)
	for i := 0; i < len(rank); i++ {
		rank[i] = 1
	}

	var find = func(n int) int {
		p := parent[n]

		// path compression
		for p != parent[p] {
			parent[p] = parent[parent[p]]
			p = parent[p]
		}

		return p
	}

	// return false if we can't union n1, n2
	var union = func(n1, n2 int) bool {
		p1, p2 := find(n1), find(n2)
		if p1 == p2 /* already connected */ {
			return false
		}

		if rank[p1] > rank[p2] {
			parent[p2] = p1
			rank[p1] += rank[p2]
		} else {
			parent[p1] = p2
			rank[p2] += rank[p1]
		}

		return true
	}
}

File: unionFind.py

# explanation: https://labuladong.github.io/algo/2/22/53/

n = 10
parent = [i for i in range(n)]
rank = [1] * n

# iterative, it won't compress path perfectly
def find(x):
    p = parent[x]
    while p != parent[x]:
        parent[x] = parent[parent[x]]
        p = parent[x]
    return p

# recursion: this is better, it'll compress all the path to O(1)
def find(x):
    if x != parent[x]:
        parent[x] = find(parent[x])
    return parent[x]

def union(x, y):
    px, py = find(x), find(y)

    if px == py: return False

    if rank[px] >= rank[py]:
        parent[py] = px
        rank[px] += rank[py]
    else:
        parent[px] = py
        rank[py] += rank[px]
    return True

def isConnected(x, y):
    return find(x) == find(y)